gateway/src/apicast/policy/soap/soap.lua

```diff
@@ -0,0 +1,127 @@
+-- This policy basically expects a SOAPAction URI in the SOAPAction header or
+-- the content-type header.
+-- The SOAPAction header is used in v1.1 of the SOAP standard:
+-- https://www.w3.org/TR/2000/NOTE-SOAP-20000508/#_Toc478383528, whereas the
+-- Content-Type header is used in v1.2 of the SOAP standard:
+-- https://www.w3.org/TR/soap12-part2/#ActionFeature
+-- The SOAPAction URI is matched against the mapping rules defined in the
+-- policy and calculates a usage based on that so it can be authorized and
+-- reported against 3scale's backend.
+
+local sub = string.sub
+local len = string.len
+local ipairs = ipairs
+local insert = table.insert
+local re = require('ngx.re')
+
+local MappingRule = require('apicast.mapping_rule')
+local Usage = require('apicast.usage')
+local mapping_rules_matcher = require('apicast.mapping_rules_matcher')
+
+local policy = require('apicast.policy')
+
+local _M = policy.new('SOAP policy')
+
+local soap_action_header = 'SOAPAction'
+local soap_action_ctype = 'application/soap+xml;'
+
+local new = _M.new
+
+local function starts_with(str, start)
+  return sub(str, 1, len(start)) == start
+end
+
+-- Extracts a SOAP action from the SOAPAction header. Returns nil when not
+-- present.
+local function soap_action_in_header(headers)
+  return headers[soap_action_header]
+end
+
+-- Extracts the SOAP action from a string that contains the parameters of a
+-- Content-Type header. The string has this format:
+-- a_param=x;action=soap_action;another_param=y
+-- This method returns the value of 'action' or nil when it's not present.
+local function soap_action_from_ctype_params(params)
+  local params_split = re.split(params, ";")
+
+  for _, param in ipairs(params_split) do
+    local header_param_split = re.split(param, "=")
+    local name = header_param_split[1]
+    local value = header_param_split[2]
+    if name == "action" then
+      return value
+    end
+  end
+
+  return nil
+end
+
+-- Extracts a SOAP action from the Content-Type header. In SOAP, the
+-- type/subtype is application/soap+xml, and the action is specified as a
+-- param in that header. When there is no SOAP action, this method returns nil.
+local function soap_action_in_ctype(headers)
+  local ctype = headers['Content-Type']
+
+  if ctype and starts_with(ctype, soap_action_ctype) then
+    local header_params = sub(ctype, len(soap_action_ctype) + 1, -1)
+    return soap_action_from_ctype_params(header_params)
+  else
+    return nil
+  end
+end
+
+-- Extracts a SOAP action URI from the SOAP Action and the Content-Type
+-- headers. When both contain a SOAP action, the Content-Type one takes
+-- precedence.
+local function extract_soap_uri()
+  local headers = ngx.req.get_headers() or {}
+  return soap_action_in_ctype(headers) or soap_action_in_header(headers)
+end
+
+local function usage_from_matching_rules(soap_action_uri, rules)
+  return mapping_rules_matcher.get_usage_from_matches(
+    nil, soap_action_uri, {}, rules)
```

Comment:
@mikz Notice that I'm sending http_method = nil here. Not sure if we should take it into account for SOAP actions.

Reply:
Hm. SOAP can be used with [GET](https://www.w3.org/TR/2007/REC-soap12-part0-20070427/#L26854) and [POST](https://www.w3.org/TR/2007/REC-soap12-part0-20070427/#L26866).
Aha! The GET does not have the `Content-Type` header. So technically it should be only POST. I guess.

But it looks like the GET is used as some REST hybrid, so I guess we could leave that to the mapping rules.
I'd possibly even go for hardcording POST there.

Reply:
I don't really know about this. The Content-Type is included in the request, not in the SOAP action URI.
Depending on what we decided here we'll need to add it to the JSON schema also. Notice that for now, I only included 'pattern', 'metric', and 'delta', and left out other fields that are present in the `proxy rules`.

Reply:
I see. We are initializing the mapping rules directly from the config. I was thinking that we could just override the `http_method` attribute with `POST` there. And then pass the real http method here. So we match it only when the request is POST. 
But it is probably not really important.